### ISSUE TITLE
SigLIP-style loss for better DDP

### ISSUE THREAD

**Issue by kddubey** [UKPLab/sentence-transformers#3119]
Hello,

[SigLIP](https://arxiv.org/abs/2303.15343) demonstrates that formulating contrastive learning as a collection of independent classifications works well.

One benefit is that the loss easily allows training with more negatives—easily b/c every summand in the loss is independent of the rest. There's no softmax/normalization like there is in MNRL. [`CachedMultipleNegativesRankingLoss`](https://sbert.net/docs/package_reference/sentence_transformer/losses.html#cachedmultiplenegativesrankingloss) already enables large batch sizes for single-device training. But for DDP, IIUC, any implementation of MNRL requires more communication overhead than SigLIP b/c of the softmax.

The SigLIP paper studies image-text data. To see if their sigmoid-style loss works for text data, I ran a [tiny experiment here](https://github.com/kddubey/blog/blob/main/sigltt/train.ipynb). It demonstrates that the performance is on par with MNRL on STS. (The notebook doesn't implement the actual, distributed SigLIP training scheme; it just checks that sigmoid instead of softmax works well for STS.)

I'm wondering if you or others are interested in incorporating a SigLIP-style training scheme into SentenceTransformers.

Thanks!

**Comment by tomaarsen** [comment]
Hello!

`CachedMultipleNegativesRankingLoss` is indeed our "solution of choice" for training with large batch sizes currently. Although discussed in #2831, it doesn't do any cross-device negatives sharing. As a result, presumably there's not a significant amount of communication overhead, right? I may be off here, though.

For example, you want to train with a batch size of 16384 and you have 8 GPUs. Each GPU can handle 64 samples at a time before OOM-ing. Then you can use DDP with 8 processes, a `per_device_train_batch_size` of 16384, and `CachedMultipleNegativesRankingLoss` with a `mini_batch_size` of 64.

The total batch size will be 131072, and that many samples are collected at the start of the step. These are then split up into 8 subbatches of 16384 samples and divided across the 8 GPUs. Each GPU will process these with CMNRL as if it's the only GPU, and the only communication is once all 8 are finished: when the gradients are averaged across all GPUs.

Please let me know if I'm overlooking something!

---

Also, looking at your script - the inputs (i.e. `scores`) must be Sigmoid-ed before they're fed through the BCELoss, right? The `scale` of 20 means that the scores range between 0 and 20, and with the bias it's between -10 and 10.

Another note: I like the idea of a dynamic scale by making it a Parameter. I experimented with this briefly, but to turn it into something "meaningful", I'd have to override the optimizer kwargs with a much higher learning rate here: https://github.com/UKPLab/sentence-transformers/blob/679ab5d38e4cf9cd73d4dcf1cda25ba2ef1ad837/sentence_transformers/trainer.py#L1206-L1219

- Tom Aarsen

**Comment by kddubey** [comment]
I think your explanation is right; CMNRL + DDP will basically do plain DDP (plus accumulating gradients w/in each device under the hood) and won't share negatives. The motivation of the issue is to enable this sharing of negatives in a DDP setup.

Here's what I mean. Say we have `G` GPUs. Our global batch size is `B`, and each anchor in our dataset has `N` negatives. Each GPU is responsible for computing the loss for its `B/G` anchors. With (C)MNRL + DDP, each anchor in each GPU only sees its local `(B/G) * (1 + N) - 1` negatives.

SigLIP is not exactly plain DDP. It allows each anchor to efficiently see all other negatives in the global batch (see Figure 1 and Section 3.3 of [SigLIP](https://arxiv.org/abs/2303.15343)), i.e., each of the `B/G` anchors in each GPU sees `B * (1 + N) - 1` negatives—`~G` times as many as before. The result is hopefully that you get a better update after each batch, which gets you a better model in less time. (Also, we should still be able to accumulate mini-batches w/ the sigmoid loss, as already done in CMNRL.)

We can achieve seeing as many negatives as you want via accumulating mini-batches in CMNRL. But that accumulation will happen in sequence instead of in parallel.

Does this motivation sound right?

Now that I'm writing this out though, I realized that SigLIP will do `G` times as many similarity calculations: (C)MNRL + DDP does `(B/G) * (B/G) * (1 + N) * G` while SigLIP does `B * B * (1 + N)`. This fact plus the discussion in #2831 makes it definitely possible that CMNRL is not slower than SigLIP. So we'd probably need to have some concrete SigLIP DDP experiments before anchoring (ha) too hard on it.

I'm now less convinced about it / will probably happily stick w/ CMNRL for now.

> Also, looking at your script - the inputs (i.e. `scores`) must be Sigmoid-ed before they're fed through the BCELoss, right?

The script uses [`BCEWithLogitsLoss`](https://pytorch.org/docs/stable/generated/torch.nn.BCEWithLogitsLoss.html), which assumes the inputs are logits. (I should've called the object `bce_with_logits_loss` instead of `bce_loss`.)

> Another note: I like the idea of a dynamic scale by making it a Parameter. I experimented with this briefly, but to turn it into something "meaningful", I'd have to override the optimizer kwargs with a much higher learning rate here:

Good to know! That's something I missed in my experiment; the `scale` and `bias` only shifted by 0.002 lol

**Comment by tomaarsen** [comment]
The motivation is indeed sound: sharing negatives allows for bigger batches, but I think it does not necessarily make sense to use cross-device negatives if we can also arbitrarily increase the batch size per device. All it really does is introduce cross-device communication, right?

> The script uses [BCEWithLogitsLoss](https://pytorch.org/docs/stable/generated/torch.nn.BCEWithLogitsLoss.html), which assumes the inputs are logits. (I should've called the object bce_with_logits_loss instead of bce_loss.)

Oops, I totally missed that. I was indeed stuck on `bce_loss`. I was already a bit surprised you got such good results while seemingly using the loss incorrectly.

---

I think there is merit in providing more elaboration in the documentation somewhere (either the Distributed Training section, or the CMNRL API reference) about really large batch size cases. I see a lot of papers use cross-device batches to try and increase their batch size, not realising that `GradCache` can solve all of their problems.

So researchers seem to look for cross-device solutions in Sentence Transformers too, not realising that we purposefully don't support it (as I think CMNRL is an equivalent or superior option). 

- Tom Aarsen

**Comment by kddubey** [comment]
> All it really does is introduce cross-device communication, right?

It does, but SigLIP's cross-device communication is relatively efficient. SigLIP shifts embeddings from each device to its neighbor ("collective permute" in the paper) instead of doing all-gathers. I believe this shift has to happen `G - 1` times per batch though. So it's definitely possible that the increase in negatives doesn't justify the cost in communication time. There would need to be hard experiments to say either way.

Overall, I think my concerns are mostly washed away. Thank you for the discussion!

**Comment by kddubey** [comment]
Another application of the sigmoid/multi-label loss is that it seamlessly supports multi-positive and multi-negative data. (Pretty sure both positives and negatives can come in variable numbers.) That way you don't need to unfold `{anchor: "...", positives: ["..."], negatives: [...]}` data into `{anchor: "...", positive: "...", negatives: [...]}` records. There are some wrinkles that I wanna iron out, e.g., correct batch-level labeling. But this seems like a use case.

**Comment by tomaarsen** [comment]
Currently, ST does not support variable number of positives/negatives as it prevents nice matrices, but I'm open to options if the results are worthwhile. I just scrolled through https://github.com/kddubey/mpnrl - some interesting work there. My understanding is that it's faster/higher GPU memory because your approach groups multiple positives per anchor?

- Tom Aarsen

**Comment by kddubey** [comment]
> My understanding is that it's faster/higher GPU memory because your approach groups multiple positives per anchor?

I think the main benefit is that it keeps batch sizes steady for datasets where the number of anchors per positives is skewed. No-duplicates sampling causes batch sizes to decay when there's high skewness [(see the first plot here)](https://github.com/kddubey/mpnrl/blob/main/compare_dataloaders.ipynb). As a result, peak memory is determined by the first few, large batches. With MPNRL, plain sampling is fine for the loss. As a result, batch sizes and memory usage are even across training.

I'll study what level of skewness is needed to see this level of decay for no-duplicates sampling, how often this skewness is encountered in real datasets, and whether MPNRL is on par w/ MNRL in terms of task performance.
